### DISCUSSION TITLE
GPUs and/or CUDA

### DISCUSSION THREAD

**Discussion by aladshaw3** [idaholab/moose#21152] [Q&A General] [answered]
I know that PETSc has some methods to utilize GPUs and was wondering if MOOSE makes use of GPUs? 

If so, how would users invoke GPUs for scaling on hybrid computer systems? 

And if not, are there any future plans for making MOOSE scalable on hybrid systems (CPUs and GPUs)? 

Thanks.

**Comment by GiudGiud** [comment]
Hello

There are plans. Though not sure if we have something laid out somewhere to share. 
@fdkong can call Petsc for MOOSE on GPU.

Guillaume

**Comment by fdkong** [comment, marked as answer]
> I know that PETSc has some methods to utilize GPUs and was wondering if MOOSE makes use of GPUs?

MOOSE can now seamlessly use GPUs for algebra operations (linear solvers, preconditioners) via PETSc. It will significantly speed up the simulation if it is expensive on the solvers, which is valid for many applications. 

We form matrix and vector on CPU and move that to GPU to solve

In the future, we plan to explore the approaches to move more calculations on GPU. It will also depend on the resources.

**Reply by aladshaw3** [reply]
Thanks for the information. As a follow up, I assume that only certain PETSc operations (specific linear solvers or preconditioners) utilize GPUs. Is there a comprehensive list somewhere so I know which solvers to invoke? Thanks.

**Reply by fdkong** [reply]
This was a great question. We (PETSc team) will work on a table for that. In general, AMG and other iterative methods work pretty well on GPUs and direct solvers such as LU, CC do not play well with GPUs because of data dependency. In theory, at the algebra side, the GPU support is almost  about Vec and Mat. Here is a roadmap at PETSc side https://petsc.org/release/overview/gpu_roadmap/

For practice,  you could start with a simple example and then use "-snes_view" to check what solvers are you using and use "-log_view" to check what operations are carried out on GPU/CPU. By doing that, you will have enough information to understand the simulation.

**Reply by mangerij** [reply]
This is a fantastic development!

**Reply by ajdubas** [reply]
Good morning,

I was wondering if you could shed a little more light on how to get MOOSE to solve GPU?  So far I have built PETSc with CUDA and have tried passing in various -dm_mat_type and -dm_vec_type flags to my MOOSE app to utilise the GPU solvers in PETSc but so far to no avail; -log_view shows no data transfer or GPU FLOPS.  Do you have a guide to what I need to do to solve on GPU?  Or any thoughts on what I might be missing?

Many Thanks in Advance,

Aleks

**Reply by GiudGiud** [reply]
@fdkong can you please share your branch where you were doing that?

**Reply by fdkong** [reply]
All GPU related stuffs have been merged into libMesh and MOOSE. You might use `-mat_type` and `-vec_type` instead of `-dm_mat_type` and `-dm_vec_type` to take GPU matrix and vector. 

In MOOSE, we do not really support DM.  BTW,  if you still have issue, please post the result of "-snes_view" and "-log_view" to us. Otherwise, we are simply blind.

Thanks for asking

**Reply by ajdubas** [reply]
Ok, my initial approach was to use -mat_type and -vec_type of aijcuda and cuda respectively; are these the right types to be using?

Cheers,

Aleks

**Reply by ajdubas** [reply]
An update: I have managed to confirm that the compiled PETSc is GPU enabled with the flags -mat_type seqaijcusparse -vec_type seqcuda passed directly to PETSc.  However, when I try the same with a MOOSE problem it does not work neither as part of the [Executioner] block nor as direct command line inputs.  Where do I go from here?

All the best,

Aleks

**Reply by GiudGiud** [reply]
Hi,

Could you please attach the run log with with  the result of "-snes_view" and "-log_view" passed to Petsc

thanks
Guillaume

**Reply by ajdubas** [reply]
Is there an example input file or command line invocation for running on GPU?

**Reply by GiudGiud** [reply]
See @fdkong 's instructions above

**Reply by GiudGiud** [reply]
@NamjaeChoi if you have some advice for @ajdubas

**Reply by ajdubas** [reply]
Ok, it seems adding -mat_type and -vec_type into getCommonPetscFlags() in framework/src/utils/PetscSupport.C enables the GPU calls to propagate to PETSc.  PR on its way.

**Reply by ljeft** [reply]
Hello@fdkong , I am trying to use GPU in moose recently. can you give me a case or documentation help?

**Reply by GiudGiud** [reply]
@fdkong

**Reply by lindsayad** [reply]
If you have a GPU enabled PETSc, then just passing `-mat_type` and `-vec_type` should work

**Reply by Traiwit** [reply]
@GiudGiud i've read this whole thread 3 times, and still not sure how to enable GPU with my app/input_file

> All GPU related stuffs have been merged into libMesh and MOOSE.

so I dont have to manually install the `GPU enabled PETSc` anymore? 

also how do I 'passing/adding'  `-mat_type and -vec_type into getCommonPetscFlags() in framework/src/utils/PetscSupport.C ` this is very confused to me. Could you please put together the instruction to common users like myself?

Thank you very much!

**Reply by Traiwit** [reply]
> also how do I 'passing/adding' -mat_type and -vec_type into getCommonPetscFlags() in framework/src/utils/PetscSupport.C 

okay I did this 

```
MultiMooseEnum
getCommonPetscFlags()
{
  return MultiMooseEnum(
      "-dm_moose_print_embedding -dm_view -ksp_converged_reason -ksp_gmres_modifiedgramschmidt "
      "-ksp_monitor -ksp_monitor_snes_lg-snes_ksp_ew -ksp_snes_ew -snes_converged_reason "
      "-snes_ksp -snes_ksp_ew -snes_linesearch_monitor -snes_mf -snes_mf_operator -snes_monitor "
      "-snes_test_display -snes_view",
      "-mat_type -vec_type",
      true);
}
```

is this all?

**Reply by GiudGiud** [reply]
No not at all these are just common options. 

You can pass any of the petsc flags

**Reply by GiudGiud** [reply]
You still need to install petsc with nvcc if you want to use cuda I m afraid. 

@grmnptr do you gave instructions somewhere?

**Reply by ljeft** [reply]
If my graphics card is AMD and I find that PETSC fails to fully support HIP (Mat), does this mean MOOSE fails to support HIP.

**Reply by lindsayad** [reply]
It's probably best if your mpi is GPU aware. You can configure mpich with `--with-hip=yes` (AMD) or `--with-cuda=yes` (nvidia). Similarly you need to configure petsc with `--with-hip=1` or `--with-cuda=1`. There are a few solvers that have full GPU support. These include PETSc's native GAMG as well as hypre's boomeramg. When you've configured your mpi and petsc for cuda/hip support and built moose on top of these, then you can run with `-mat_type aijcusparse -vec_type cuda` or with `-mat_type aijhipsparse -vec_type hip` with solver types like those I just mentioned. Note that you cannot use these GPU mat types for the preconditioning matrix and use matrix-free finite differencing for the Jacobian, so you cannot use PJFNK, you must use NEWTON.

**Reply by DanoLV** [reply]
Hello. Can you tell me exactly how to build PETSc with nvcc? 
I have been trying to build MOOSE with CUDA aware OpenMPI an PETSc, but my efforts have been unsuccessful until now.
OpenMPI is CUDa aware in my build, but when I try to build PETSc it fails with SUPERLU_DIST.

![image](https://github.com/idaholab/moose/assets/59875578/301e6459-f18a-4ea9-b9ec-1bacff63fa1c)
[Uploading configure.log…]()

**Reply by milljm** [reply]
@DanoLV  Please create a new discussion thread. You may link this discussion in your new discussion thread if you wish. But it is poor practice to ask for help for your issue, when another is also asking for help in the same thread.

**Reply by milljm** [reply]
And this goes for everyone; This thread has already been answered. It cannot have another. If this thread does not answer your question, please create a _new_ question using a new discussion, so that it may have a _new_ answer.

**Reply by tayebel** [reply]
hello, am new to moose . and am trying to apply GPU acceleration, any manual doc can help please? thanks

**Reply by milljm** [reply]
We do not have instructions on how to achieve this. But there is a PR which contains a work in progress set of configs we are passing to PETSc you can make use of. Specifically this file in the following PR:

https://github.com/idaholab/moose/pull/28114/files#diff-20b6a3ebe3b74d550bfa94a728056daf5fc0c18acc92d5653fe40898384d7fa9

These PETSc configure options are tacked on to our normal PETSc build script:

https://github.com/idaholab/moose/blob/8ccd0043313d608c3a8bfb23b2bc0327771a00dd/apptainer/petsc.def#L66

If I am reading everything correctly, that would mean you want to do something like:

```bash
cd moose/scripts
./update_and_rebuild_petsc.sh --download-kokkos \
  --download-kokkos-kernels \
  --with-cuda \
  --with-cuda-arch=70 \
  --download-slate
```
_you will also need to do everything necessary to be in a good environment with a working compiler before running the above. Perhaps follow one of our manual installation instructions (don't do `conda create -n moose moose-dev`)_

Not sure how much effort you want to put into building your own CUDA PETSc stack, but the above link would be a start.

**Reply by lindsayad** [reply]
This is also predicated on building mpi with something like `--with-cuda`

**Reply by tayebel** [reply]
Somehow when I run : ./update_and_rebuild_petsc.sh --download-kokkos \
  --download-kokkos-kernels \
  --with-cuda \
  --with-cuda-arch=70 \
  --download-slate
  
  it is not detecting CUDA compiler in my path, I tried to pass it manually and am getting the same issue. I am using [Windows Subsystem for Linux] installation. Is a docker image installation will work for me ?

**Reply by milljm** [reply]
WSL2 (probably the version you are using), should work. I have done it before... Though IIRC it was very finicky dependent on the paths, and what I _needed NOT to install_ using `apt-get`. Let me see if I can find the link. It was on NVIDIA's forum somewhere...

Edit:
Here we go: https://docs.nvidia.com/cuda/wsl-user-guide/index.html

That bit in [section 3](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#cuda-support-for-wsl-2), there is an annotation about "users must not install any NVIDIA GPU Linux driver within WSL". -that was my issue the first time.

You might find this link useful as well, as I recall I had to peruse here and ask some questions myself: https://gitlab.com/petsc/petsc/-/issues/1646
